numpy/_core/src/umath/loops_trigonometric.dispatch.cpp

```diff
@@ -108,162 +111,143 @@ simd_sincos_f32(const float *src, npy_intp ssrc, float *dst, npy_intp sdst,
-        if (ssrc == 1) {
-            x_in = hn::LoadN(f32, src, len);
-        } else {
-            x_in = hn::GatherIndexN(f32, src, src_index, len);
-        }
-        opmask_t nnan_mask = hn::Not(hn::IsNaN(x_in));
-        // Eliminate NaN to avoid FP invalid exception
-        x_in = hn::IfThenElse(nnan_mask, x_in, zerosf);
-        opmask_t simd_mask = hn::Le(hn::Abs(x_in), max_cody);
-        /*
-         * For elements outside of this range, Cody-Waite's range reduction
-         * becomes inaccurate and we will call libc to compute cosine for
-         * these numbers
-         */
-        if (!hn::AllFalse(f32, simd_mask)) {
-            vec_f32 x = hn::IfThenElse(hn::And(nnan_mask, simd_mask), x_in, zerosf);
-
-            vec_f32 quadrant = hn::Mul(x, two_over_pi);
-            // round to nearest, -0.0f -> +0.0f, and |a| must be <= 0x1.0p+22
-            quadrant = hn::Add(quadrant, rint_cvt_magic);
-            quadrant = hn::Sub(quadrant, rint_cvt_magic);
-
-            // Cody-Waite's range reduction algorithm
-            vec_f32 reduced_x = simd_range_reduction_f32(
-                x, quadrant, codyw_pio2_highf, codyw_pio2_medf, codyw_pio2_lowf
-            );
-            vec_f32 reduced_x2 = hn::Mul(reduced_x, reduced_x);
-
-            // compute cosine and sine
-            vec_f32 cos = simd_cosine_poly_f32(reduced_x2);
-            vec_f32 sin = simd_sine_poly_f32(reduced_x, reduced_x2);
-
-            vec_s32 iquadrant = hn::NearestInt(quadrant);
-            if (trig_op == SIMD_COMPUTE_COS) {
-                iquadrant = hn::Add(iquadrant, ones);
-            }
-            // blend sin and cos based on the quadrant
-            opmask_t sine_mask = hn::RebindMask(f32, hn::Eq(hn::And(iquadrant, ones), hn::Zero(s32)));
-            cos = hn::IfThenElse(sine_mask, sin, cos);
-
-            // multiply by -1 for appropriate elements
-            opmask_t negate_mask = hn::RebindMask(f32, hn::Eq(hn::And(iquadrant, twos), twos));
-            cos = hn::MaskedSubOr(cos, negate_mask, zerosf, cos);
-            cos = hn::IfThenElse(nnan_mask, cos, hn::Set(f32, NPY_NANF));
-
-            if (sdst == 1) {
-                hn::StoreN(cos, f32, dst, len);
-            } else {
-                hn::ScatterIndexN(cos, f32, dst, dst_index, len);
-            }
-        }
-        if (!hn::AllTrue(f32, simd_mask)) {
-            npy_uint64 simd_maski;
-            hn::StoreMaskBits(f32, simd_mask, (uint8_t*)&simd_maski);
-            float NPY_DECL_ALIGNED(NPY_SIMD_WIDTH) ip_fback[hn::Lanes(f32)];
-            hn::Store(x_in, f32, ip_fback);
-
-            // process elements using libc for large elements
-            if (trig_op == SIMD_COMPUTE_COS) {
-                for (unsigned i = 0; i < hn::Lanes(f32); ++i) {
-                    if ((simd_maski >> i) & 1) {
-                        continue;
+    hn::Transform1(
+            f32, src, ssrc, dst, sdst, len,
+            [&](vec_f32 &x_in) -> vec_f32 HWY_ATTR {
```

Comment:
This is a fairly large function. Rather than defining it inline, can we add an explicit function with an appropriate name to make it more readable?

Reply:
Can we decide whether we want to do that in a future PR? It'd require moving all of the predefined constants around whilst this PR simply exchanges the loop for the wrapper.

Reply:
Is it not as simple as moving all the `const` variables out of the loop and marking them `static const` at the top of the file?

Reply:
I misinterpreted what you were asking, I thought you wanted to move the constants closer to their relevant functions rather than just put them at the top of the file. I can take a look, I did start moving stuff into the relevant functions though, so will see if I can clean up that way 😸